Add docs for exchange materialization (#15193)
prestodb/presto: presto-docs/src/main/sphinx/admin/exchange-materialization.rst

========================
Exchange Materialization
========================

Presto allows exchange materialization to support memory intensive queries.
This mechanism brings MapReduce-style execution to Presto's MPP architecture runtime,
and can be applied together with :doc:`/admin/spill`.

Introduction
------------

As with other MPP databases, Presto leverages RPC shuffle to achieve efficient and
low-latency query execution for join and aggregation. However, RPC shuffle
also requires all the producers and receivers have to be executed concurrently until the
query finished.

To illustrates this, consider the aggregation query:

.. code-block:: sql

    SELECT custkey, SUM(totalprice)
    FROM orders
    GROUP BY custkey

The following figure demonstrates how this query executes in Presto classic mode:

.. figure:: ../images/rpc_shuffle_execution.png
   :align: center

With exchange materialization, the intermediate shuffle data is written to disk (currently,

> it is always a temporary Hive bucketed table). This opens opportunity for flexible scheduling policies

Reviewer: ```suggestion
it is always a temporary Hive bucketed table). This creates an opportunity for flexible scheduling policies
```

Author: @aweisberg : What about `opens the opportunity for XXX`?

Reviewer: That WFM

on the aggregation side, as only a subset of aggregation data needs to be held in memory at the
same time -- this execution strategy is called "grouped execution" in Presto.

> .. figure:: ../images/materialized_shuffle_execution.png

Reviewer: Capitalize first letter of bullet points?

"Schedule fewer partitions concurrently to reduce memory requirements"

   :align: center

Using Exchange Materialization
------------------------------

Exchange materialization can be enabled on per-query basis by setting the following 3 session properties:
``exchange_materialization_strategy``, ``partitioning_provider_catalog`` and ``hash_partition_count``:

.. code-block:: sql

    SET SESSION exchange_materialization_strategy='ALL';

    -- Set partitioning_provider_catalog to the Hive connector catalog
    SET SESSION partitioning_provider_catalog='hive';

>     -- We recommend to set hash_partition_count to be at least 5X-10X about the number of cluster size

Reviewer: ```suggestion
    -- We recommend setting hash_partition_count to be at least 5X-10X cluster size
```

>     -- We recommend setting hash_partition_count to be at least 5X-10X about the number of cluster size

Reviewer: ```suggestion
    -- We recommend setting hash_partition_count to be at least 5X-10X the cluster size
```
